File: limablu_mobile/lib/services/wallet.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

class WalletEntry {
  final DateTime ts;
  final String type; // "topup" | "deduct"
  final double kwh;
  final double amountKsh;
  final String? ref;

  WalletEntry({
    required this.ts,
    required this.type,
    required this.kwh,
    required this.amountKsh,
    this.ref,
  });

  Map<String, dynamic> toJson() => {
        "ts": ts.toIso8601String(),
        "type": type,
        "kwh": kwh,
        "amount_ksh": amountKsh,
        "ref": ref,
      };

  static WalletEntry fromJson(Map<String, dynamic> m) => WalletEntry(
        ts: DateTime.parse(m["ts"] as String),
        type: (m["type"] as String),
        kwh: (m["kwh"] as num).toDouble(),
        amountKsh: (m["amount_ksh"] as num).toDouble(),
        ref: (m["ref"] as String?),
      );
}

class WalletService {
  static const _storageKey = "wallet_ledger_v1";
  List<WalletEntry> _entries = [];

  List<WalletEntry> get entries => List.unmodifiable(_entries);

  double get balanceKwh =>
      _entries.fold(0.0, (s, e) => s + (e.type == "topup" ? e.kwh : -e.kwh));

  Future<void> load() async {
    final prefs = await SharedPreferences.getInstance();
    final s = prefs.getString(_storageKey);
    if (s == null) {
      _entries = [];
      return;
    }
    final list = (jsonDecode(s) as List).cast<Map<String, dynamic>>();
    _entries = list.map(WalletEntry.fromJson).toList()
      ..sort((a, b) => b.ts.compareTo(a.ts));
  }

  Future<void> _save() async {
    final prefs = await SharedPreferences.getInstance();
    final s = jsonEncode(_entries.map((e) => e.toJson()).toList());
    await prefs.setString(_storageKey, s);
  }

  Future<void> addTopUp({
    required double kwh,
    required double amountKsh,
    required String reference,
  }) async {
    _entries.insert(
      0,
      WalletEntry(
        ts: DateTime.now(),
        type: "topup",
        kwh: kwh,
        amountKsh: amountKsh,
        ref: reference,
      ),
    );
    await _save();
  }

  Future<void> addDeduction({required double kwh, String? reason}) async {
    _entries.insert(
      0,
      WalletEntry(
        ts: DateTime.now(),
        type: "deduct",
        kwh: kwh,
        amountKsh: 0,
        ref: reason,
      ),
    );
    await _save();
  }

  Future<void> clearAll() async {
    _entries.clear();
    await _save();
  }
}

File: emappa-mobile/src/services/wallet.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { WalletEntry } from '../types';

const STORAGE_KEY = 'wallet_ledger_v1';

class WalletService {
  private entries: WalletEntry[] = [];

  get allEntries(): WalletEntry[] {
    return [...this.entries];
  }

  get balanceKwh(): number {
    return this.entries.reduce((sum, e) => {
      return sum + (e.type === 'topup' ? e.kwh : -e.kwh);
    }, 0);
  }

  async load(): Promise<void> {
    try {
      const raw = await AsyncStorage.getItem(STORAGE_KEY);
      if (raw) {
        const list = JSON.parse(raw) as WalletEntry[];
        this.entries = list.sort(
          (a, b) => new Date(b.ts).getTime() - new Date(a.ts).getTime()
        );
      } else {
        this.entries = [];
      }
    } catch (e) {
      console.error('Failed to load wallet:', e);
      this.entries = [];
    }
  }

  private async save(): Promise<void> {
    try {
      await AsyncStorage.setItem(STORAGE_KEY, JSON.stringify(this.entries));
    } catch (e) {
      console.error('Failed to save wallet:', e);
    }
  }

  async addTopUp(kwh: number, amountKsh: number, reference: string): Promise<void> {
    const entry: WalletEntry = {
      ts: new Date().toISOString(),
      type: 'topup',
      kwh,
      amount_ksh: amountKsh,
      ref: reference,
    };
    this.entries.unshift(entry);
    await this.save();
  }

  async addDeduction(kwh: number, reason?: string): Promise<void> {
    const entry: WalletEntry = {
      ts: new Date().toISOString(),
      type: 'deduct',
      kwh,
      amount_ksh: 0,
      ref: reason,
    };
    this.entries.unshift(entry);
    await this.save();
  }

  async clearAll(): Promise<void> {
    this.entries = [];
    await this.save();
  }
}

// Singleton instance
export const walletService = new WalletService();
